madara-alliance/madara
# feat(client): l1<>l2 messaging crate

# Pull Request type

- Feature

## What is the current behavior?

Resolves: #204

## What is the new behavior?

- Added a crate with a worker that listen for LogMessageToL2 event on main contract and handle it as the old madara did.
- Added some entries to the db for messaging related data

## Does this introduce a breaking change?

No 

## Usefull information

🚨Require #208 🚨
This PR include a lot of code from #208 as we started from it to build with the feature from the eth crate.

A few notes on the current design:
- There is no guarantee that L1->L2 messages will be executed in a centralized context.
- There is no ordering guarantee, nonces are solely used 

## Current Progress
### Overall Workflow
- [X] Create a stream of LogMessageToL2 events
- [X] Get last synced block from Messaging DB
- [X] Consume the stream and log event
- [ ] Process message 
    - [X] Parse tx fee
    - [X] Parse transaction from event
    - [X] Check if message has already been processed
    - [X] Build transaction
    - [Waiting for mempool #198 ] Submit tx to mempool
    - [X] Update Messaging DB
    
### Specific case to handle    
- [ ] Handle Message Cancellation (https://github.com/keep-starknet-strange/madara/issues/1638)

### Tests

- [ ] E2E test 1
    - Launch Anvil Node
    - Launch Worker
    - Send L1->L2 message
    - Assert that event is emitted on L1
    - Assert that even was caught by the worker with correct data
    - Assert the tx hash computed by the worker is correct
    - Assert that the tx has been included in the mempool
    - Assert that DB was correctly updated (last synced block & nonce)
    - Assert that the tx was correctly executed

- [ ]  E2E test 2
    - Should fail if we try to send multiple messages with same nonces

- [ ]  E2E test 3 
    - Message Cancellation
